projects/com.oracle.truffle.llvm.asm.amd64/src/com/oracle/truffle/llvm/asm/amd64/AsmFactory.java

```diff
@@ -393,114 +453,368 @@ LLVMInlineAssemblyRootNode finishInline() {
     }
 
+    void setPrefix(String prefix) {
+        this.currentPrefix = prefix;
+    }
+
+    void createInt(AsmImmediateOperand nr) {
+        long id = nr.getValue();
+        if (id == 3) {
+            statements.add(new LLVMAMD64BreakpointNode(sourceSection));
+        } else {
+            statements.add(new LLVMI32UnsupportedInlineAssemblerNode(null));
+        }
+    }
+
     void createOperation(String operation) {
         switch (operation) {
             case "clc":
             case "cld":
             case "cli":
             case "cmc":
                 statements.add(new LLVMI32UnsupportedInlineAssemblerNode(null));
-                return;
+                break;
             case "lahf": {
-                LLVMExpressionNode lahf = LLVMAMD64LahfNodeGen.create(getFlag(LLVMAMD64Flags.CF), getFlag(LLVMAMD64Flags.PF), getFlag(LLVMAMD64Flags.ZF), getFlag(LLVMAMD64Flags.SF));
+                LLVMExpressionNode lahf = LLVMAMD64LahfNodeGen.create(getFlag(LLVMAMD64Flags.CF), getFlag(LLVMAMD64Flags.PF), getFlag(LLVMAMD64Flags.AF), getFlag(LLVMAMD64Flags.ZF),
+                                getFlag(LLVMAMD64Flags.SF));
                 LLVMExpressionNode write = getOperandStore(PrimitiveType.I8, new AsmRegisterOperand("ah"), lahf);
                 statements.add(write);
                 break;
             }
-            case "popfw":
-            case "pushfw":
+            case "popf":
+            case "popfw": {
+                LLVMExpressionNode read = LLVMAMD64PopwNodeGen.create();
+                LLVMExpressionNode write = LLVMAMD64WriteFlagswNodeGen.create(getFlagWrite(LLVMAMD64Flags.CF), getFlagWrite(LLVMAMD64Flags.PF), getFlagWrite(LLVMAMD64Flags.AF),
+                                getFlagWrite(LLVMAMD64Flags.ZF), getFlagWrite(LLVMAMD64Flags.SF), getFlagWrite(LLVMAMD64Flags.OF), read);
+                statements.add(write);
+                break;
+            }
+            case "pushf":
+            case "pushfw": {
+                LLVMExpressionNode flags = LLVMAMD64ReadFlagswNodeGen.create(getFlag(LLVMAMD64Flags.CF), getFlag(LLVMAMD64Flags.PF), getFlag(LLVMAMD64Flags.AF), getFlag(LLVMAMD64Flags.ZF),
+                                getFlag(LLVMAMD64Flags.SF), getFlag(LLVMAMD64Flags.OF));
+                LLVMExpressionNode write = LLVMAMD64PushwNodeGen.create(flags);
+                statements.add(write);
+                break;
+            }
             case "sahf":
             case "stc":
             case "std":
             case "sti":
                 statements.add(new LLVMI32UnsupportedInlineAssemblerNode(null));
-                return;
+                break;
             case "nop":
+                if ("rep".equals(currentPrefix)) {
+                    // TODO: pause
+                }
+                break;
+            case "hlt":
+            case "mfence":
+            case "lfance":
+            case "sfence":
+                // TODO: implement properly
```

Comment:
can we also throw an IllegalStateException / fail here instead of a silent TODO?

Reply:
There is already a `LLVMFence` class that we could reuse to implement these nodes. However, we assumed that we start off by making inline assembly work for single-threaded execution. I think we currently also ignore the `lock` prefix that makes instructions execute automatically.